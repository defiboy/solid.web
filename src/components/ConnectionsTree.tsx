import React from 'react';

import { Action, ActionCreator } from "redux";

import styled from 'styled-components';
import { Icon, Tree, Menu } from 'antd';

import { Connection, Contract } from "../redux/types";

const DirectoryTree = Tree.DirectoryTree;

const { TreeNode } = Tree;
const MenuItem = Menu.Item;

const SidebarHeader = styled.div`
    display: flex;
    align-items: center;
    justify-content: space-between;
`

const SidebarTitle = styled.h3`
    color: white;
    font-weight: 100;
    padding: 0;
    margin: 0;
    font-size: 1em;
`

const SidebarHeaderButtons = styled.div`

`

const TreeNodeStyled = styled(TreeNode)`
    span {
        color: white;
    }
    .
`

const MenuStyled = styled(Menu)`
    z-index: 1000;
    border: none !important;
`
const MenuItemStyled = styled(MenuItem)`
    height: 2em !important;
    line-height: 2em !important;
    margin-top: 0 !important;
    margin-bottom: 0 !important;
   
    &:hover{
        background: #25b864;
        color: white;
    }
`

interface Props {
    connections: Connection[]
    contracts: Contract[]
    onNewConnectionClick: ActionCreator<Action>
    onContractSelected: ActionCreator<Action>
}

interface State {
    rightClickNodeTreeItem: any
    selectedKeys: any[]
}
// : React.FC<Props> = ({ connections, onNewConnectionClick, contracts }: Props) => {
export class ConnectionsTree extends React.Component<Props, State> {

    constructor(props: Props) {
        super(props);
        this.state = {
            rightClickNodeTreeItem: {},
            selectedKeys: []
        }
    }

    onSelect = (selectedKeys: any, info: any) => {
        const contractToShow = this.props.contracts.find((item) => {
            return item._id === selectedKeys[0];
        })
        this.props.onContractSelected(contractToShow);
        this.setState({
            selectedKeys
        });
    };

    rightClickOnTree = ({ event, node }: any) => {
        const id = node.props['eventKey'];
        this.setState({
            rightClickNodeTreeItem: {
                pageX: event.pageX,
                pageY: event.pageY,
                id: id,
                categoryName: node.props['eventKey']
            },
            selectedKeys: [id]
        });
    }

    getNodeTreeRightClickMenu = () => {
        const { pageX, pageY } = { ...this.state.rightClickNodeTreeItem } as any
        if (!pageX || !pageY) {
            return (<div></div>);
        }
        const menu = (<MenuStyled style={{ position: 'absolute', left: `${pageX}px`, top: `${pageY}px` }}>
            <MenuItemStyled key='1'>Deploy</MenuItemStyled>
            <MenuItemStyled key='2'>Open Console</MenuItemStyled>
        </MenuStyled>);
        return (menu)
    }

    render() {
        const { connections, contracts, onNewConnectionClick } = this.props;
        return (
            <div style={{ overflow: "scroll", height: "100%" }}>
                {this.getNodeTreeRightClickMenu()}
                < SidebarHeader >
                    <SidebarTitle>Connections</SidebarTitle>
                    <SidebarHeaderButtons>
                        <Icon type="plus" style={{ color: "white", paddingRight: "0.5em" }} onClick={onNewConnectionClick} />
                        <Icon type="down" style={{ color: "white" }} />
                    </SidebarHeaderButtons>
                </SidebarHeader >
                {connections && connections.length > 0 && <DirectoryTree
                    onSelect={this.onSelect}
                    multiple
                    onRightClick={this.rightClickOnTree}
                    selectedKeys={this.state.selectedKeys}
                    defaultExpandAll style={{ color: "white" }}>
                    {connections.map((item) => {
                        return <TreeNodeStyled icon={<Icon type="database" />} title={item.name} key={item.url} style={{ color: "white" }}>
                            <TreeNodeStyled title="Contract Instances" key="0-0-0" style={{ color: "white" }}>
                                {contracts && contracts.length > 0 && contracts.map((contract: Contract) => {
                                    return <TreeNodeStyled title={contract.name} key={contract._id} isLeaf />
                                })}
                            </TreeNodeStyled>
                        </TreeNodeStyled>
                    })}
                </DirectoryTree>}
            </div >
        )
    }
}