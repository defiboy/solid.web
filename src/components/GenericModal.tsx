import React from 'react'
import * as R from 'ramda'

import { Modal, Button } from 'antd'
import { Formik, FormikErrors, FormikProps } from 'formik'

interface Props<FormFields> {
  FormComponent: React.ComponentType<FormComponentProps<FormFields>>
  initialValues: FormFields
  visible: boolean
  onCancel: () => void
  loading: boolean
  validator: (item: FormFields) => FormikErrors<FormFields>
  onSubmit: (item: FormFields) => void
  title: string
  disableSubmitButton?: boolean
  buttonText?: string
  formId: string;
}

type FormComponentProps<FormFields> = { fields: FormFields } & Handlers

type OnChangeHandler<FormFields> = <K extends keyof FormFields>(s: K, a: FormFields[K]) => void

interface Handlers {
  onSubmit: ((event?: React.FormEvent<HTMLFormElement> | undefined) => void)
}

interface State<FormFields> {
  fields: FormFields
}

export class GenericModal<FormFields> extends React.Component<Props<FormFields>, State<FormFields>> {
  constructor(props: Props<FormFields>) {
    super(props)
    this.state = { fields: props.initialValues }
  }

  onChange: OnChangeHandler<FormFields> = (field, value) => {
    const result = R.merge(this.state.fields, { [field]: value }) as any
    this.setState({
      fields: result
    })
  }

  public render() {
    const { FormComponent, visible, onCancel, loading, validator, title, disableSubmitButton, formId, onSubmit } = this.props
    const { fields } = this.state
    return (
      <Modal
        width={500}
        visible={visible}
        title={title}
        onCancel={onCancel}
        footer={[
          <Button data-testid={`cancel-button-${formId}`} key="cancel" onClick={onCancel}>
            Cancel
          </Button>,
          <Button
            data-testid={`submit-button-${formId}`}
            form={formId}
            key="submit"
            htmlType="submit"
            type="primary"
            disabled={disableSubmitButton || false}
            loading={loading}>
            {this.props.buttonText || 'Save'}
          </Button>
        ]}
      >
        <div>
          <Formik
            initialValues={fields}
            onSubmit={onSubmit}
            enableReinitialize={true}
            validate={validator}
            render={({ handleSubmit }: FormikProps<FormFields>) => (
              <FormComponent onSubmit={handleSubmit} fields={fields} />
            )}
          />

        </div>
      </Modal>
    )
  }
}
