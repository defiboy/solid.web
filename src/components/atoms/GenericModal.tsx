import React from "react"
import * as R from "ramda";

import { Modal, Button } from "antd";
import { Formik, FormikErrors, FormikProps } from "formik";

const FORM_TITLE = "ItemForm";

interface Props<FormFields> {
    FormComponent: React.ComponentType<FormComponentProps<FormFields>>;
    initialValues: FormFields;
    visible: boolean;
    onCancel: () => void
    loading: boolean;
    validator: (item: FormFields) => FormikErrors<FormFields>
    onSubmit: (item: FormFields) => void // TODO
    title: string
    disableSubmitButton?: boolean
    buttonText?: string
}

type FormComponentProps<FormFields> = { fields: FormFields } & Handlers<
    FormFields
>;

type OnChangeHandler<FormFields> = <K extends keyof FormFields>(
    s: K,
    a: FormFields[K]
) => void;

interface Handlers<FormFields> {
    onSubmit: (e?: React.FormEvent<HTMLFormElement> | undefined) => void
}

interface State<FormFields> {
    fields: FormFields;
}

class GenericModal<FormFields> extends React.Component<Props<FormFields>, State<FormFields>> {
    constructor(props: Props<FormFields>) {
        super(props);
        this.state = { fields: props.initialValues };
    }

    public onChange: OnChangeHandler<FormFields> = (field, value) => {
        const result = R.merge(this.state.fields, { [field]: value }) as any
        this.setState({
            fields: result
        });
    };

    submit = (values: FormFields) => {
        console.log("Submitting", values)
        this.props.onSubmit(values);
    }

    public render() {
        const { FormComponent, visible, onCancel, loading, validator, title, disableSubmitButton } = this.props;
        const { fields } = this.state;
        return (

            <Modal
                width={500}
                visible={visible}
                title={title} // TODO change title if it is edit or creation mode
                onCancel={onCancel}
                footer={[<Button key="cancel" onClick={onCancel}>Cancel</Button>,
                <Button
                    form={FORM_TITLE}
                    key="submit"
                    htmlType="submit"
                    type="primary"
                    disabled={disableSubmitButton || false}
                    loading={loading}>{this.props.buttonText || "Save"}</Button>]}>
                <div>
                    <Formik
                        initialValues={fields}
                        onSubmit={this.submit}
                        enableReinitialize={true}
                        validate={validator}
                        render={({
                            handleSubmit
                        }: FormikProps<FormFields>) => (
                                <FormComponent onSubmit={handleSubmit} fields={fields} />
                            )}
                    />
                </div>

            </Modal>)
    }
}
export default GenericModal;

