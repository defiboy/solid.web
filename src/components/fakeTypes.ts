export interface GenericType {
    name: string
    url: string
}
