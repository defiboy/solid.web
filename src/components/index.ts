export { GenericModal } from './GenericModal'
export { TextAreaFormItem } from './TextAreaFormItem'
export { InputFormItem } from './InputFormItem'
