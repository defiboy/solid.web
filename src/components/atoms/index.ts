import GenericModal from "./GenericModal";
export { GenericModal }
export { TextAreaField } from "./TextAreaField"
export { TextField } from "./TextField"

