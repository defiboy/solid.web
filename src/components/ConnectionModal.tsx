import React from 'react'
import { bindActionCreators, Dispatch, ActionCreator, Action } from 'redux'
import { connect } from 'react-redux'

import { Form } from 'antd'
import { FormikErrors, Field } from 'formik'

import { Connection, Status } from '../redux/types'
import { ApplicationState } from '../redux/reducers'
import { createOrUpdateConnection, createConnectionCancelled } from '../redux/actions'
import { GenericModal, TextField } from './atoms'

const FORM_TITLE = 'ItemForm' // TODO change to dinamic

interface Props {
  visible: boolean
  loading: boolean
  itemToEdit?: Connection
  createOrUpdateConnection: (item: Connection) => void
  createConnectionCancelled: ActionCreator<Action>
  // onCancel: () => void
}

export class ConnectionModalComponent extends GenericModal<Connection> {}

export class ConnectionModal extends React.Component<Props> {
  saveConnection = (item: Connection) => {
    // if (this.state.itemToEdit) { // TODO
    //     // update values
    // }
    console.log('save connection', item)
    this.props.createOrUpdateConnection(item)
  }

  render() {
    return (
      <ConnectionModalComponent
        title="Add Connection"
        onSubmit={this.saveConnection}
        visible={this.props.visible}
        loading={this.props.loading}
        onCancel={this.props.createConnectionCancelled}
        initialValues={{ name: '13', url: '', transactionReceipts: [] }}
        validator={(items: Connection) => {
          const errors: FormikErrors<Connection> = {}
          if (!items.name) {
            errors.name = 'Required'
          }
          if (!items.url) {
            errors.url = 'Required'
          }
          return errors
        }}
        FormComponent={({ fields: { name, url }, onSubmit }) => (
          <Form id={FORM_TITLE} onSubmit={onSubmit}>
            <Field
              name="name"
              render={(innerProps: any) => <TextField {...innerProps} label="Name" placeHolder="http://" />}
            />
            <Field
              name="url"
              render={(innerProps: any) => (
                <TextField {...innerProps} label="Blockchain URL" placeHolder="JSON RPC endpoint" />
              )}
            />
          </Form>
        )}
      />
    )
  }
}

const mapStateToProps = (state: ApplicationState) => {
  return {
    createConnection: state.appState.createConnection,
    visible: state.appState.createConnection.status === Status.Started,
    submitted: state.appState.createConnection.status === Status.Completed,
    loading: state.appState.createConnection.status === Status.InProgress
  }
}

const mapDispatchToProps = (dispatch: Dispatch) => {
  return bindActionCreators(
    {
      createOrUpdateConnection,
      createConnectionCancelled
    },
    dispatch
  )
}

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(ConnectionModal)
