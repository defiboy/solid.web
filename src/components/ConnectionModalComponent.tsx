import { GenericModal } from './atoms'
import { Connection } from '../redux/types'

export class ConnectionModalComponent extends GenericModal<Connection> { }
