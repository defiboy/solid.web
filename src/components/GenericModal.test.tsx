import React from 'react';
import { FormikErrors, Field, FieldProps } from 'formik';
import { render } from '@testing-library/react'
import { Form } from 'antd'

import { InputFormItem } from 'components'

import { GenericType } from './fakeTypes'
import { GenericModal } from './GenericModal'

class GenericModalComponent extends GenericModal<GenericType> { }

describe('GenericModal', () => {

  const MODAL_TITLE = "modal title"
  const FORM_ID = "modalid"

  const onSubmitMockHandler = jest.fn()
  const onCancelMockHandler = jest.fn()

  const initialValues: GenericType = {
    name: '',
    url: ''
  }

  const renderWithGenericModal = () => {
    return render(<GenericModalComponent
      formId={FORM_ID}
      title={MODAL_TITLE}
      onCancel={onCancelMockHandler}
      onSubmit={onSubmitMockHandler}
      initialValues={initialValues}
      visible={true}
      loading={true}
      validator={(values: GenericType): FormikErrors<GenericType> => {
        let errors: FormikErrors<GenericType> = {};
        if (!values.name) {
          errors.name = 'Required';
        }
        if (!values.url) {
          errors.url = 'Invalid email address';
        }
        return errors;
      }}
      FormComponent={({ onSubmit }) => (
        <Form id={FORM_ID} onSubmit={onSubmit}>
          <Field
            name="name"
            render={(innerProps: FieldProps) => <InputFormItem {...innerProps} label="Name" placeHolder="Name" />}
          />
          <Field
            name="url"
            render={(innerProps: FieldProps) => <InputFormItem {...innerProps} label="Url" placeHolder="Url" />}
          />
        </Form>
      )}
    >
    </GenericModalComponent>)
  }

  it('renders a modal with basic elements', () => {
    const { getByLabelText, getByTestId } = renderWithGenericModal()

    expect(getByLabelText(`${MODAL_TITLE}`)).toBeInTheDocument()
    expect(getByTestId('input-name')).toBeInTheDocument()
    expect(getByTestId('input-url')).toBeInTheDocument()
    expect(getByTestId(`submit-button-${FORM_ID}`)).toBeInTheDocument()
    expect(getByTestId(`cancel-button-${FORM_ID}`)).toBeInTheDocument()
  })

  it('renders a modal', () => {
    const { baseElement, debug } = renderWithGenericModal()

    debug(baseElement)
    expect(baseElement).toMatchSnapshot()
  })
})
