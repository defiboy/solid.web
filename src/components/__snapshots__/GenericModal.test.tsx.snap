// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GenericModal renders a modal 1`] = `
<body
  style="overflow: hidden;"
>
  <div />
  <div>
    <div>
      <div
        class="ant-modal-mask fade-appear"
      />
      <div
        aria-labelledby="rcDialogTitle1"
        class="ant-modal-wrap "
        role="dialog"
        tabindex="-1"
      >
        <div
          class="ant-modal zoom-appear"
          role="document"
          style="width: 500px;"
        >
          <div
            aria-hidden="true"
            style="width: 0px; height: 0px; overflow: hidden;"
            tabindex="0"
          />
          <div
            class="ant-modal-content"
          >
            <button
              aria-label="Close"
              class="ant-modal-close"
              type="button"
            >
              <span
                class="ant-modal-close-x"
              >
                <i
                  aria-label="icon: close"
                  class="anticon anticon-close ant-modal-close-icon"
                >
                  <svg
                    aria-hidden="true"
                    class=""
                    data-icon="close"
                    fill="currentColor"
                    focusable="false"
                    height="1em"
                    viewBox="64 64 896 896"
                    width="1em"
                  >
                    <path
                      d="M563.8 512l262.5-312.9c4.4-5.2.7-13.1-6.1-13.1h-79.8c-4.7 0-9.2 2.1-12.3 5.7L511.6 449.8 295.1 191.7c-3-3.6-7.5-5.7-12.3-5.7H203c-6.8 0-10.5 7.9-6.1 13.1L459.4 512 196.9 824.9A7.95 7.95 0 0 0 203 838h79.8c4.7 0 9.2-2.1 12.3-5.7l216.5-258.1 216.5 258.1c3 3.6 7.5 5.7 12.3 5.7h79.8c6.8 0 10.5-7.9 6.1-13.1L563.8 512z"
                    />
                  </svg>
                </i>
              </span>
            </button>
            <div
              class="ant-modal-header"
            >
              <div
                class="ant-modal-title"
                id="rcDialogTitle1"
              >
                modal title
              </div>
            </div>
            <div
              class="ant-modal-body"
            >
              <div>
                <form
                  class="ant-form ant-form-horizontal"
                  id="modalid"
                >
                  <div
                    class="ant-row ant-form-item"
                  >
                    <div
                      class="ant-col ant-form-item-label"
                    >
                      <label
                        class=""
                        for="name"
                        title="Name"
                      >
                        Name
                      </label>
                    </div>
                    <div
                      class="ant-col ant-form-item-control-wrapper"
                    >
                      <div
                        class="ant-form-item-control"
                      >
                        <span
                          class="ant-form-item-children"
                        >
                          <input
                            class="ant-input ant-input-lg"
                            data-testid="input-name"
                            id="name"
                            name="name"
                            placeholder="Name"
                            type="text"
                            value=""
                          />
                        </span>
                      </div>
                    </div>
                  </div>
                  <div
                    class="ant-row ant-form-item"
                  >
                    <div
                      class="ant-col ant-form-item-label"
                    >
                      <label
                        class=""
                        for="url"
                        title="Url"
                      >
                        Url
                      </label>
                    </div>
                    <div
                      class="ant-col ant-form-item-control-wrapper"
                    >
                      <div
                        class="ant-form-item-control"
                      >
                        <span
                          class="ant-form-item-children"
                        >
                          <input
                            class="ant-input ant-input-lg"
                            data-testid="input-url"
                            id="url"
                            name="url"
                            placeholder="Url"
                            type="text"
                            value=""
                          />
                        </span>
                      </div>
                    </div>
                  </div>
                </form>
              </div>
            </div>
            <div
              class="ant-modal-footer"
            >
              <button
                class="ant-btn"
                data-testid="cancel-button-modalid"
                type="button"
              >
                <span>
                  Cancel
                </span>
              </button>
              <button
                class="ant-btn ant-btn-primary ant-btn-loading"
                data-testid="submit-button-modalid"
                form="modalid"
                type="submit"
              >
                <i
                  aria-label="icon: loading"
                  class="anticon anticon-loading"
                >
                  <svg
                    aria-hidden="true"
                    class="anticon-spin"
                    data-icon="loading"
                    fill="currentColor"
                    focusable="false"
                    height="1em"
                    viewBox="0 0 1024 1024"
                    width="1em"
                  >
                    <path
                      d="M988 548c-19.9 0-36-16.1-36-36 0-59.4-11.6-117-34.6-171.3a440.45 440.45 0 0 0-94.3-139.9 437.71 437.71 0 0 0-139.9-94.3C629 83.6 571.4 72 512 72c-19.9 0-36-16.1-36-36s16.1-36 36-36c69.1 0 136.2 13.5 199.3 40.3C772.3 66 827 103 874 150c47 47 83.9 101.8 109.7 162.7 26.7 63.1 40.2 130.2 40.2 199.3.1 19.9-16 36-35.9 36z"
                    />
                  </svg>
                </i>
                <span>
                  Save
                </span>
              </button>
            </div>
          </div>
          <div
            aria-hidden="true"
            style="width: 0px; height: 0px; overflow: hidden;"
            tabindex="0"
          />
        </div>
      </div>
    </div>
  </div>
</body>
`;
